### ISSUE TITLE
Support attaching additional databases to a dqlite connection

### ISSUE THREAD

**Issue by cole-miller** [canonical/dqlite#441]
Right now, it seems a lot of dqlite code (down to the VFS layer) assumes that operations on a single leader connection only affect the database that was `OPEN`ed on that connection. But SQLite has an [in-band mechanism](https://www.sqlite.org/lang_attach.html) for "attaching" additional database files to an open connection. We've disabled this for now (#440), but we might want to provide proper support for it, particularly since `ATTACH` is used internally by SQLite in the implementation of `VACUUM` (see #435) and possibly other things. I'm opening this issue so we have a place to discuss what that support might look like and how big of a project it might be (we can close this if we decide that `ATTACH` isn't something we want to support).

**Comment by freeekanayaka** [comment]
It can be probably be made a no-op indeed.

However, I think it'd be worth asking the Juju team why do they need the ability to attach databases in the first place. Note that in WAL mode transactions across attached databases are **not** atomic.

I'm not saying that attaching databases is a feature dqlite should not support, but I'm not entirely sure it's really the most appropriate approach for Juju.

**Comment by freeekanayaka** [comment]
The problem mentioned in the PR thread seems to be that following the procedure described in the SQLite docs you point does not work for us, because VACUUM will attach a temporary database behind the curtains (IIUC).

However, my assumptions are that 1) the Juju team is asking for deletion (not reset) 2) with some care, we can implement deletion at the VFS level without using the procedure described in the SQLite docs.

In other words, I don't think we need attach support in order to implement database deletion. I believe having a dedicated FSM command plus a vfs function that drops all the pages from memory (or from disk, if disk mode is enabled) is enough. Some care should be taken to ensure that no open client connection exists against the database being deleted, but that should be about it.

**Comment by freeekanayaka** [comment]
For the "normal" dqlite in-memory mode snapshots should be synchronous so the problem described in the PR should not be an issue (please correct me if that was changed recently, and snapshots are async also in the in-memory mode). For disk-mode the async snapshot logic should be already be prepared to have the underlying database change concurrently, so deletion of the entire database is "just" a concurrent change. I'm not entirely sure what is needed, just saying that it should be possible to support a snapshots that run concurrently to a delete-database FSM command.

Regarding the garbage-collection idea in that PR, let's always remember that whatever we do, FSM must be fully deterministic. So the concerns that @cole-miller expresses [here](https://github.com/canonical/dqlite/pull/431#issuecomment-1318769966) and I express [here](https://github.com/canonical/dqlite/pull/431#issuecomment-1318547103) are valid.

Anyway, as said I believe we should first investigate the feasibility of having snapshots happen concurrently with deletion, which should be possible with some care.

BTW: @MathieuBordere is disk-mode already being used in real-world? At least as far as we know.

**Comment by MathieuBordere** [comment]
> BTW: @MathieuBordere is disk-mode already being used in real-world? At least as far as we know.

Real world no, but there's a lot of intention to use it in the real world at some point not too far away in the future.

This might be (is) a discussion for another place, but I think the "disk-mode" is one of the ways to solve the problem of snapshotting a fairly large dqlite database. The problem is that if a database is several GB's in size and there are quite a lot of write transactions going on, then dqlite will very frequently dump the whole database in the form of a snapshot to disk, which leads to taking up a lot of disk IO bandwidth.  disk-mode could offer a sort of disk-based fsm where after a checkpoint (when the still in-memory WAL is flushed to disk) we more or less have a snapshot of the fsm already on disk, thus not requiring us to dump several GB's of data time and time again to disk. Disk-based fsm's are discussed in paragraph 5.2 of the [raft dissertation](https://raw.githubusercontent.com/ongardie/dissertation/master/book.pdf).

**Comment by freeekanayaka** [comment]
For cases where the dqlite FSM is not going to be too large and users opt for in-memory mode, we can surely think about tweaking the snapshot frequency and possibly "throttle" disk writes when persisting snapshots, so that part does not have too much impact on the system.
